### PR TITLE
[Core] Added support for submission-time task names.

### LINKED ISSUE #10371: [Core] Submission-time task names.
### Feature Request

Allow the user to specify submission-time task names, such that the task name is used as the process title (for displaying in the dashboard and debugging) and possibly other areas (e.g. core worker stats). Using the task function name as the user-friendly identifier results in poor UX when many tasks are executed using the same remote function, especially when there exists a semantically meaningful name that could be attached to the task at submission time.

### API and Semantics

The user can specify a task name at submission time:
```python
@ray.remote
def foo(x):
    return x + 1

foo.options(name="my_special_foo").remote(1)
```
Unlike Ray actor names, this would **not** "detach" the task, and global uniqueness of these task names would **not** be enforced. Given the existing semantics for `actor.options(name="foo").remote()`, I'm imagining that this new task name option would only apply for actor and non-actor tasks, not for actor creation.

When viewing the machine view of the Ray dashboard, the current task (in the workers/PID column) for a given core worker would be given by this task name, if set. The task name could also be included in other metrics/stats payloads, such as the core worker stats.

### Possible Implementation

I think that this semantically meaningful task name should be easy to propagate if there's a dedicated field on the `TaskSpec`. If the user supplies a `name` at task submission time, the `task_name` field is set to that; if not, it's set to the task function's name, the current behavior.

Flow - setting field in task spec + using said field to set proctitle:
1. Add a `task_name` field to the `TaskSpec` [class](https://github.com/ray-project/ray/blob/0178d6318ead643fff6d9ed873c88f71fdac52e7/src/ray/common/task/task_spec.h#L30-L216)/[proto](https://github.com/ray-project/ray/blob/fe6daef85e1f3cd2aef12b39487124a6e2dcc2b9/src/ray/protobuf/common.proto#L147-L189) and to the [`TaskOptions` class](https://github.com/ray-project/ray/blob/0178d6318ead643fff6d9ed873c88f71fdac52e7/src/ray/core_worker/common.h#L54-L64).
2. Properly set it on the task spec in [`CoreWorker::SubmitTask`](https://github.com/ray-project/ray/blob/0aec4cbccb5862a7808f98b90696236da9e2a03e/src/ray/core_worker/core_worker.cc#L1257-L1287) and [`CoreWorker::SubmitActorTask`](https://github.com/ray-project/ray/blob/0aec4cbccb5862a7808f98b90696236da9e2a03e/src/ray/core_worker/core_worker.cc#L1398-L1436) to either the user-provided `task_name` if given or set it to the function's name if not (current behavior).
3. Pull the `task_name` off of the `TaskSpec` class in [`CoreWorker::ExecuteTask`](https://github.com/ray-project/ray/blob/0aec4cbccb5862a7808f98b90696236da9e2a03e/src/ray/core_worker/core_worker.cc#L1635-L1757) and pass to language-specific task execution handler.
4. Use said `task_name` to [set the proctitle in the language-specific task execution handlers](https://github.com/ray-project/ray/blob/0aec4cbccb5862a7808f98b90696236da9e2a03e/python/ray/_raylet.pyx#L389).

### Concerns/Open Questions

1. Should the submission-time task name be appended or prepended to the function name, yielding something like `some_module.some_func:{task_name}` or `{task_name}:some_module.some_func`, instead of replacing the function name? Or should the user be expected to craft such a task name in their code and use _that_ as the `.options(name=task_name)` argument if that's what they want? I'm thinking the latter, but having to do that crafting for actor methods may be annoying.
2. Will this be confusing given the existence of named actors and the already established semantics thereof? Should a similar concept be provided for actors, and if so, how could we delineate between detached actors and named actors?

### PR DESCRIPTION
**Pull request by clarkzinzow**
## Why are these changes needed?

Allows the user to specify a submission-time name for a task, providing better dashboard and debugging UX for remote functions that are reused with differing semantics. This PR adds a `name` option to `options()` for normal and actor tasks, and adds a corresponding `name` field to the task spec. If `name` is given, it is set as the task name. If `name` is not given, the task name defaults to a name derived from the function descriptor, preserving the current behavior.  This task name is used as the process title in the Python worker (therefore showing up in the dashboard) and is set in the core worker stats via the `current_task_name` field.

Actor creation, which already has a `name` option, is mostly excluded from this PR (although that name is now included in the actor creation task).

## Related issue number

Closes #10371 

## TODO

- [x] Add tests.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: cpp/src/ray/runtime/abstract_ray_runtime.cc
Comment (clarkzinzow) [inline cpp/src/ray/runtime/abstract_ray_runtime.cc:72]: Not going to expose this until there's something akin to an `AbstractRayRuntime::Options()` API.

File: src/ray/core_worker/core_worker.cc
Comment (ericl) [inline src/ray/core_worker/core_worker.cc:1425]: It's probably a small overhead, but I wonder if we should just leave the task name as empty and try to derive it as before, instead of computing it and saving it as part of the descriptor.

Perhaps try `ray microbenchmark` on a large machine and see if there's any measurable degradation in performance?

Comment (clarkzinzow) [reply src/ray/core_worker/core_worker.cc:1425]: I thought about going that route, but I felt like it made the task spec abstraction messier. If we know the task name at submission time (either user-provided or derived from the function descriptor), why not set it on the task spec? If we set it, all downstream consumers of the task spec will know that `task_spec.GetName()` will give them a non-empty user-friendly task name, which is a clean contract. If we don't set it, all downstream consumers will have to know that the task name field might be empty, in which case the consumer is expected to construct a user-friendly task name off of the function descriptor, which seems less clean and will duplicate that logic at every point in which the task name is consumed.

If something downstream wanted to be able to delineate between a user-provided task name and a task name derived from the function descriptor (e.g. so they can construct a task name that makes the most sense for their context that's different than the default one generated by the function descriptor), _that_ I could see as an argument to leave the task name empty if not given by the user. Can you see any cases like that popping up?

Comment (clarkzinzow) [reply src/ray/core_worker/core_worker.cc:1425]: Another option, if the uncommon case arises in which the default task name isn't what a consumer wants when the user doesn't specify a task name, would be to move this logic to `TaskSpec::GetName()`: if `!name_.empty()`, return `name_`, otherwise return `GetFunctionDescriptor()->DefaultTaskName()`. That would make `SubmitTask` and `SubmitActorTask` cleaner, and we could expose something like `TaskSpec::IsNameUserProvided() { return !name_.empty(); }` so consumers can delineate between user-specified and default task names. Each language that wraps the `TaskSpecification` proto message would have to implement this logic, but given that this is an internal proto message, that seems reasonable.

I'm hoping to preserve the happy common path, where most task spec consumers are don't care whether the task name was user-provided or the default derived from the function descriptor: they're just happy to have a human-readable task name.

**Comment by ericl** [comment]
I think it's fine, if it doesn't show up on microbenchmarks. Performance
was my main concern there, I agree it is cleaner to do it the current way
in the PR.

File: src/ray/common/task/task_spec.h
Comment (rkooo567) [inline src/ray/common/task/task_spec.h:139]:
```suggestion
  const std::string GetName() const;
```

File: src/ray/common/task/task_spec.cc
Comment (rkooo567) [inline src/ray/common/task/task_spec.cc:195]:
```suggestion
const std::string TaskSpecification::GetName() const { return message_->name(); }
```
? Any reason why the return value is not const?

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:1317]:
```suggestion
  const auto actor_name = actor_creation_options.name;
```

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:1318]: same here?

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:1424]:
```suggestion
  const auto task_name = task_options.name.empty()
```

File: src/ray/core_worker/lib/java/jni_utils.h
Comment (rkooo567) [inline src/ray/core_worker/lib/java/jni_utils.h:533]: nice!

**Review by rkooo567 — COMMENTED** [review]
Can you add a small python test to make sure it works?

File: java/api/src/main/java/io/ray/api/call/BaseTaskCaller.java
Comment (ffbin) [reply java/api/src/main/java/io/ray/api/call/BaseTaskCaller.java:21]: Ok, i will review the Java part of this PR.

**Comment by ffbin** [comment]
@clarkzinzow Can you add a java testcase, thanks.

File: src/ray/common/function_descriptor.h
Comment (clarkzinzow) [reply src/ray/common/function_descriptor.h:135]: @rkooo567 Could you elaborate on your reasoning for this change? I broke out `class_name` and `function_name` since this seemed more readable than
```c++
return typed_message_->class_name().empty() ? typed_message_->function_name() : typed_message_->class_name() + "." + typed_message_->function_name();
```
and would eliminate two unnecessary proto message accessor function calls. What you've suggested seems to keep the `class_name` and `function_name` variables, but still duplicates the proto message accessor function calls.

File: src/ray/common/task/task_spec.cc
Comment (clarkzinzow) [reply src/ray/common/task/task_spec.cc:195]: Nope, I forgot to make it const, good catch!

File: src/ray/core_worker/core_worker.cc
Comment (clarkzinzow) [inline src/ray/core_worker/core_worker.cc:1318]:
```suggestion
  const auto task_name =
```

File: doc/source/advanced.rst
Comment (clarkzinzow) [reply doc/source/advanced.rst:63]: @rkooo567 I'm happy to change this assertion to something more sensical, but what do you mean by "insert an image"?

File: python/ray/actor.py
Comment (clarkzinzow) [reply python/ray/actor.py:103]: This should support `num_returns` as well, I'll add it to the example in order to make that clear.

File: doc/source/advanced.rst
Comment (rkooo567) [reply doc/source/advanced.rst:63]: Ah, so I thought showing ray.get doesn't really have any meaning in this context (since people already knew how this works). Instead, to demonstrate what the name is doing, we can insert an image of dashboard of ps to show that the assigned name is properly set on each process.

File: python/ray/actor.py
Comment (rkooo567) [reply python/ray/actor.py:103]: If you are adding this please add a Python test for this too!

File: src/ray/common/function_descriptor.h
Comment (rkooo567) [reply src/ray/common/function_descriptor.h:135]: oh in this case, I meant to delete the first two lines (it seems like it is same for other methods). Sorry for confusion. Also, this comment is not a hard requirement. (I initially thought it can save one copy constructor, but behavior might be the same. I am not 100% sure... cpp world :)).

**Comment by rkooo567** [comment]
@clarkzinzow Not sure if you have the access, but can you remove the `author-action-required ` label once you addressed all comments? If you don't have access, just ping us in the comment! (I will remove it for you).

File: doc/source/advanced.rst
Comment (clarkzinzow) [reply doc/source/advanced.rst:63]: Ah, ok! I was planning on adding `assert setproctitle.getproctitle() == "ray::special_f"` within the task to demonstrate that the process name will be properly set. Would you prefer this, a dashboard image, or both?

File: src/ray/common/function_descriptor.h
Comment (clarkzinzow) [reply src/ray/common/function_descriptor.h:135]: Ah that makes more sense! I'll double-check that C++ behavior.

Comment (rkooo567) [reply src/ray/common/function_descriptor.h:135]: Yeah I am not sure if it is true because I am also not a cpp expert. I think it can be actually just the same. Let me know what's the right behavior.

File: doc/source/advanced.rst
Comment (rkooo567) [reply doc/source/advanced.rst:63]: I think both will be great in this case. If I have to choose one, I prefer the image because the behavior is more obvious in my opinion! (But you might have better opinion here since you are an actual user!)

**Comment by clarkzinzow** [comment]
@ericl Pertinent microbenchmark results:

Before this PR:
```
single client tasks sync per second 661.23 +- 1.39
single client tasks async per second 5264.85 +- 92.92
multi client tasks async per second 5743.71 +- 410.95
1:1 actor calls sync per second 1110.19 +- 9.57
1:1 actor calls async per second 3940.26 +- 96.0
1:1 actor calls concurrent per second 3830.18 +- 46.79
1:n actor calls async per second 6792.23 +- 121.44
n:n actor calls async per second 7198.73 +- 162.59
n:n actor calls with arg async per second 2887.41 +- 22.62
1:1 async-actor calls sync per second 853.31 +- 2.12
1:1 async-actor calls async per second 2508.33 +- 22.15
1:1 async-actor calls with args async per second 1630.4 +- 17.11
1:n async-actor calls async per second 4574.6 +- 18.54
n:n async-actor calls async per second 4760.74 +- 25.85
```

This PR:
```
single client tasks sync per second 665.39 +- 0.85
single client tasks async per second 5496.23 +- 88.58
multi client tasks async per second 6072.32 +- 36.59
1:1 actor calls sync per second 1139.73 +- 6.47
1:1 actor calls async per second 3936.75 +- 52.74
1:1 actor calls concurrent per second 3777.86 +- 68.22
1:n actor calls async per second 6834.45 +- 158.89
n:n actor calls async per second 7301.09 +- 29.36
n:n actor calls with arg async per second 2908.99 +- 22.46
1:1 async-actor calls sync per second 859.59 +- 1.12
1:1 async-actor calls async per second 2494.17 +- 8.23
1:1 async-actor calls with args async per second 1634.08 +- 9.63
1:n async-actor calls async per second 4468.12 +- 38.24
n:n async-actor calls async per second 4839.3 +- 13.27
```

No obvious performance regressions.

**Review by ericl — APPROVED** [review]
Lgtm then

**Comment by ffbin** [comment]
Hi @clarkzinzow , the java CI job failed and is related to the JNI code.

**Comment by clarkzinzow** [comment]
@ffbin The Java CI job 8 (with env `JDK='Oracle JDK 8'`) is failing with the same error as the most recent master builds:

- [This PR](https://travis-ci.com/github/ray-project/ray/jobs/380113617)
- [Recent master build](https://travis-ci.com/github/ray-project/ray/jobs/380111345)

I can look into that, but I think that failure is orthogonal to this PR.

As for the streaming test suite that's failing with a JNI error, it seems like that's orthogonal to this PR as well. Recent master builds are failing with that same error in the wheel testing/building build:

- [This PR's streaming test suite job](https://travis-ci.com/github/ray-project/ray/jobs/380113618)
- [Recent master wheel testing/building job](https://travis-ci.com/github/ray-project/ray/jobs/380111349)

I've tried running the Java tests locally with Bazel pointing at a couple of different JDKs (8 and 11), and have gotten everything to compile, but at runtime I get (unrelated) segfaults on one and many failed test configurations on the other: `java.lang.NoClassDefFoundError: Could not initialize class java.lang.management.ManagementFactory`. So I have two questions here:

1. Do you have any tips for getting my local Java development environment working with the Ray repo? I haven't been able to get Bazel's out-of-the-box Java toolchain to work.
2. Which Java CI jobs are supposed to be succeeding? The results of recent master CI jobs suggest that the Ray Java API/runtime test suite and the Ray streaming test suite are both consistently failing in Travis right now.

File: src/ray/common/function_descriptor.h
Comment (clarkzinzow) [reply src/ray/common/function_descriptor.h:135]: We can keep the readability and the elimination of the two extra proto message accessor function calls by changing the `std::string`s to const references, which is what is returned by the [C++ generated proto code for message field accessors](https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#proto3_string). Only the necessary copies, in either returning a copy of `function_name` or constructing the concatenated class name + function name, should be made.

**Comment by ffbin** [comment]
@clarkzinzow The steps to run Java testcase on mac are as follows:
1 ./ build.sh -l java
2 cd java
3 mvn clean install -DskipTests - Dcheckstyle.skip=true
4 mvn test -pl test -Dtest="xxx" - Dcheckstyle.skip

**Comment by clarkzinzow** [comment]
@ffbin Hmm ok, I thought that building and testing purely through Bazel would work. I just tried that out, looks like an issue with a preconditions check:
```
ERROR: /mnt/data/workspace/ray/streaming/java/BUILD.bazel:94:19: Building streaming/java/libio_ray_ray_streaming-runtime-class.jar (75 source files, 2 source jars) failed (Exit 1): java failed: error executing command
  (cd /home/clark/.cache/bazel/_bazel_clark/4b0c7d95a1a2a417d84b6788f088ac1e/execroot/com_github_ray_project_ray && \
  exec env - \
    LC_CTYPE=en_US.UTF-8 \
    PATH='/home/clark/.kubectx:~/downloads/istio-1.4.2/bin:/usr/local/go/bin:/home/clark/go/bin:/home/clark/.cargo/bin:/home/clark/bin:/home/clark/.local/bin:/usr/local/go/bin:/home/clark/go/bin:/home/clark/.cargo/bin:/home/clark/.kubectx:~/downloads/istio-1.4.2/bin:/home/clark/bin:/home/clark/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/clark/.fzf/bin' \
  external/remotejdk11_linux/bin/java -XX:+UseParallelOldGC -XX:-CompactStrings '--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED' '--add-exports=
jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED' '--add-opens=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED' '--patch-module=java.compiler=external/remote_java_tools_linux/java_tools/java_compiler.jar' '--patch-module=jdk.compiler=external/remote_java_tools_linux/java_tools/jdk_compiler.jar' '--add-opens=java.base/java.nio=ALL-UNNAMED' '--add-opens=java.base/java.lang=ALL-UNNAMED' -jar external/remote_java_tools_linux/java_tools/JavaBuilder_deploy.jar @bazel-out/k8-opt/bin/stream
ing/java/libio_ray_ray_streaming-runtime-class.jar-0.params @bazel-out/k8-opt/bin/streaming/java/libio_ray_ray_streaming-runtime-class.jar-1.params)
Execution platform: @local_config_platform//:host
streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/transfer/ChannelId.java:145: error: [PreconditionsInvalidPlaceholder] Preconditions only accepts the %s placeholder in error message strings
        "fromTaskId %d is larger than %d", fromTaskId, Short.MAX_VALUE);
        ^
    (see https://errorprone.info/bugpattern/PreconditionsInvalidPlaceholder)
  Did you mean '"fromTaskId %s is larger than %s", fromTaskId, Short.MAX_VALUE);'?
streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/transfer/ChannelId.java:147: error: [PreconditionsInvalidPlaceholder] Preconditions only accepts the %s placeholder in error message strings
        "toTaskId %d is larger than %d", fromTaskId, Short.MAX_VALUE);
        ^
    (see https://errorprone.info/bugpattern/PreconditionsInvalidPlaceholder)
  Did you mean '"toTaskId %s is larger than %s", fromTaskId, Short.MAX_VALUE);'?
Target //java:ray_java_pkg failed to build
INFO: Elapsed time: 58.261s, Critical Path: 41.41s
INFO: 39 processes: 32 linux-sandbox, 2 local, 5 worker.
FAILED: Build did NOT complete successfully
```
This is [documented](https://guava.dev/releases/20.0/api/docs/com/google/common/base/Preconditions.html) and should probably be fixed, I wonder how this hasn't been hit before. 🤔 

>Master Java CI job should be fixed yesterday, and you can rebase master.

As I said above, the most recent master builds have failing Java CI jobs. E.g. here's the [most recent completed Travis build](https://github.com/ray-project/ray/runs/1059284758), where each of the Java builds/test suites are failing with the same errors as the builds for this PR. And this PR was rebased onto master less than 5 hours ago, onto commit 3b10b67a15bb07cf3f918126dd3c5a7047b467df.

File: python/ray/tests/test_advanced_3.py
Comment (rkooo567) [inline python/ray/tests/test_advanced_3.py:667]: Can you use fstring instead? We dropped 3.5 support, so we can use fstring.

File: python/ray/tests/test_advanced_3.py
Comment (rkooo567) [inline python/ray/tests/test_advanced_3.py:684]: same here

File: python/ray/_raylet.pyx
Comment (rkooo567) [inline python/ray/_raylet.pyx:389]: Please use fstring

**Review by rkooo567 — APPROVED** [review]
LGTM as long as @ffbin signs off for Java. Just the last batch comments for fstring.

File: python/ray/tests/test_advanced_3.py
Comment (clarkzinzow) [reply python/ray/tests/test_advanced_3.py:667]: Oh awesome, I didn't know that y'all dropped 3.5 support! I'm the biggest fstring advocate you'll ever meet, I'll get on that.

**Comment by ffbin** [comment]
> @ffbin Hmm ok, I thought that building and testing purely through Bazel would work. I just tried out building via `./build.sh`, looks like an issue with a preconditions check:
> 
> ```
> ERROR: /mnt/data/workspace/ray/streaming/java/BUILD.bazel:94:19: Building streaming/java/libio_ray_ray_streaming-runtime-class.jar (75 source files, 2 source jars) failed (Exit 1): java failed: error executing command
>   (cd /home/clark/.cache/bazel/_bazel_clark/4b0c7d95a1a2a417d84b6788f088ac1e/execroot/com_github_ray_project_ray && \
>   exec env - \
>     LC_CTYPE=en_US.UTF-8 \
>     PATH='/home/clark/.kubectx:~/downloads/istio-1.4.2/bin:/usr/local/go/bin:/home/clark/go/bin:/home/clark/.cargo/bin:/home/clark/bin:/home/clark/.local/bin:/usr/local/go/bin:/home/clark/go/bin:/home/clark/.cargo/bin:/home/clark/.kubectx:~/downloads/istio-1.4.2/bin:/home/clark/bin:/home/clark/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/clark/.fzf/bin' \
>   external/remotejdk11_linux/bin/java -XX:+UseParallelOldGC -XX:-CompactStrings '--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED' '--add-exports=
> jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED' '--add-opens=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED' '--patch-module=java.compiler=external/remote_java_tools_linux/java_tools/java_compiler.jar' '--patch-module=jdk.compiler=external/remote_java_tools_linux/java_tools/jdk_compiler.jar' '--add-opens=java.base/java.nio=ALL-UNNAMED' '--add-opens=java.base/java.lang=ALL-UNNAMED' -jar external/remote_java_tools_linux/java_tools/JavaBuilder_deploy.jar @bazel-out/k8-opt/bin/stream
> ing/java/libio_ray_ray_streaming-runtime-class.jar-0.params @bazel-out/k8-opt/bin/streaming/java/libio_ray_ray_streaming-runtime-class.jar-1.params)
> Execution platform: @local_config_platform//:host
> streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/transfer/ChannelId.java:145: error: [PreconditionsInvalidPlaceholder] Preconditions only accepts the %s placeholder in error message strings
>         "fromTaskId %d is larger than %d", fromTaskId, Short.MAX_VALUE);
>         ^
>     (see https://errorprone.info/bugpattern/PreconditionsInvalidPlaceholder)
>   Did you mean '"fromTaskId %s is larger than %s", fromTaskId, Short.MAX_VALUE);'?
> streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/transfer/ChannelId.java:147: error: [PreconditionsInvalidPlaceholder] Preconditions only accepts the %s placeholder in error message strings
>         "toTaskId %d is larger than %d", fromTaskId, Short.MAX_VALUE);
>         ^
>     (see https://errorprone.info/bugpattern/PreconditionsInvalidPlaceholder)
>   Did you mean '"toTaskId %s is larger than %s", fromTaskId, Short.MAX_VALUE);'?
> Target //java:ray_java_pkg failed to build
> INFO: Elapsed time: 58.261s, Critical Path: 41.41s
> INFO: 39 processes: 32 linux-sandbox, 2 local, 5 worker.
> FAILED: Build did NOT complete successfully
> ```
> 
> This is [documented](https://guava.dev/releases/20.0/api/docs/com/google/common/base/Preconditions.html) and should probably be fixed, I wonder how this hasn't been hit before. 🤔
> 
> > Master Java CI job should be fixed yesterday, and you can rebase master.
> 
> As I said above, the most recent master builds have failing Java CI jobs. E.g. here's the [most recently completed Travis build](https://github.com/ray-project/ray/runs/1059284758), where each of the Java builds/test suites are failing with the same errors as the builds for this PR. And this PR was rebased onto master less than 5 hours ago, onto commit [3b10b67](https://github.com/ray-project/ray/commit/3b10b67a15bb07cf3f918126dd3c5a7047b467df).

Hi @chaokunyang , can you help take a look, thanks.

**Review by ffbin — APPROVED** [review]
LGTM

**Comment by chaokunyang** [comment]
@clarkzinzow Can we hold this PR for serveral hours? The java ci is broken. See https://travis-ci.com/github/ray-project/ray/jobs/380156102. I'm looking itno it. Probably will be resolved today.

**Comment by rkooo567** [comment]
@clarkzinzow @chaokunyang Let me know once everything's resolved. I will merge the PR!

**Comment by rkooo567** [comment]
@chaokunyang What's the status on the master build?

**Comment by clarkzinzow** [comment]
@rkooo567 If we determine that the failing Java builds are orthogonal to this PR (which I think they are since the Java builds on current master are failing in a similar way), can we merge this and address the failing Java builds in a separate PR?

**Comment by chaokunyang** [comment]
@clarkzinzow Could you merge master? We just have some java fix merged.

**Comment by clarkzinzow** [comment]
@chaokunyang I rebased the PR onto the Java linting fix commit.

**Comment by clarkzinzow** [comment]
I just rebased onto [most recent master](https://github.com/ray-project/ray/commit/2597b56f488be1c489af819a3a23d63c49eb13aa); hopefully the streaming tests should run smoothly now. Still waiting on the [PR](https://github.com/ray-project/ray/pull/10524) fixing the `loadConfigFromGCS` Checkstyle violation to be merged. I'll check the result of this build in ~6 hours; if someone with merge permissions thinks that it looks good in the meantime, please feel free to merge it.

**Comment by rkooo567** [comment]
@clarkzinzow I will merge it tomorrow morning if the test result looks fine. I'd try to be careful because our branch cut will be in 2 days, and I'd like to avoid the PR is breaking Java builds (although it should be most likely be fine if the same tests fail on the master).

**Comment by clarkzinzow** [comment]
@rkooo567 @chaokunyang The only failures occurring in the last build are the same Java build failures that happen in the [corresponding master commit](https://github.com/ray-project/ray/commit/2597b56f488be1c489af819a3a23d63c49eb13aa). I'll rebase onto most recent master and will probably see a successful completion of the Java API and runtime test suites (streaming tests will still fail).

**Comment by rkooo567** [comment]
@clarkzinzow Will check the test result and merge it!

**Pull request merged**
